lib/pure/rational.nim

```diff
@@ -0,0 +1,228 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2015 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+
+## This module implements rational numbers, consisting of a numerator `num` and
+## a denominator `den`, both of type int. The denominator can not be 0.
+
+import math
+
+type Rational* = tuple[num, den: int]
+  ## a rational number, consisting of a numerator and denominator
+
+proc toRational*(x: SomeInteger): Rational =
+  ## Convert some integer `x` to a rational number.
+  result.num = x
+  result.den = 1
+
+proc toFloat*(x: Rational): float =
+  ## Convert a rational number `x` to a float.
+  float(x.num) / float(x.den)
+
+proc toInt*(x: Rational): int =
+  ## Convert a rational number `x` to an int. Conversion rounds if `x` does not
+  ## contain an integer value.
+  round(toFloat(x))
+
+proc reduce*(x: var Rational) =
+  ## Reduce rational `x`.
+  let common = gcd(x.num, x.den)
+  if x.den > 0:
+    x.num = x.num div common
+    x.den = x.den div common
+  elif x.den < 0:
+    x.num = -x.num div common
+    x.den = -x.den div common
+  else:
+    raise newException(DivByZeroError, "division by zero")
+
+proc `+` *(x, y: Rational): Rational =
+  ## Add two rational numbers.
+  let common = lcm(x.den, y.den)
+  result.num = common div x.den * x.num + common div y.den * y.num
+  result.den = common
+  reduce(result)
+
+proc `+` *(x: Rational, y: int): Rational =
+  ## Add rational `x` to int `y`.
+  result.num = x.num + y * x.den
+  result.den = x.den
+
```

Comment:
You can define `+` using the `+=` and avoid all that code duplication:

``` nimrod
proc `+` *(x: Rational, y: Rational | int): Rational =
  result = x
  result += y

proc `+` *(x: int, y: Rational): Rational =
  result = y
  result += x
  # or result = y + x (but one more level of function calls if not inlined)
```
Similarly for  `*`, etc.